portal-impl/src/com/liferay/portlet/internal/MutablePortletParametersBase.java

```diff
@@ -0,0 +1,198 @@
+
+	public MutablePortletParametersBase(Map<String, String[]> parameterMap) {
+		super(parameterMap);
+	}
+
+	@Override
+	public MutablePortletParameters add(PortletParameters portletParameters) {
+		MutablePortletParameters oldMutablePortletParameters = clone();
+		Set<String> newParameterNames = portletParameters.getNames();
+		Map<String, String[]> parameterMap = getParameterMap();
+
+		for (String newParameterName : newParameterNames) {
+			String[] values = portletParameters.getValues(newParameterName);
+
+			String[] copiedValues = values.clone();
+
+			parameterMap.put(newParameterName, copiedValues);
+		}
+
+		_mutated = true;
+
+		return oldMutablePortletParameters;
+	}
+
+	@Override
+	public void clear() {
+		Map<String, String[]> parameterMap = getParameterMap();
+
+		parameterMap.clear();
+
+		_mutated = true;
+	}
+
+	@Override
+	public boolean isMutated() {
+		return _mutated;
+	}
+
+	@Override
+	public boolean removeParameter(String name) {
+		if (name == null) {
+			throw new IllegalArgumentException();
+		}
+
+		Map<String, String[]> parameterMap = getParameterMap();
+
+		if (parameterMap.containsKey(name)) {
+			parameterMap.remove(name);
+
+			_mutated = true;
+
+			return true;
+		}
+
+		return false;
+	}
+
+	@Override
+	public MutablePortletParameters set(PortletParameters portletParameters) {
+		MutablePortletParameters oldMutablePortletParameters = clone();
+
+		Set<String> oldParameterNames = oldMutablePortletParameters.getNames();
+
+		Set<String> newParameterNames = portletParameters.getNames();
+		Map<String, String[]> parameterMap = getParameterMap();
+
+		for (String oldParameterName : oldParameterNames) {
+			if (!newParameterNames.contains(oldParameterName)) {
+				parameterMap.remove(oldParameterName);
+
+				_mutated = true;
+			}
+		}
+
+		for (String newParameterName : newParameterNames) {
+			String[] values = portletParameters.getValues(newParameterName);
+
+			String[] copiedValues = values.clone();
+
+			parameterMap.put(newParameterName, copiedValues);
+
+			_mutated = true;
+		}
+
+		return oldMutablePortletParameters;
+	}
+
+	@Override
+	public String setValue(String name, String value) {
+		String[] oldValues = setValues(name, new String[] {value});
+
+		_mutated = true;
+
+		if ((oldValues != null) && (oldValues.length > 0)) {
+			return oldValues[0];
+		}
+
+		return null;
+	}
+
+	@Override
+	public String setValue(String name, String value, boolean append) {
+		String[] oldValues;
+
+		if (value == null) {
+			oldValues = setValues(name, null, append);
+		}
+		else {
+			oldValues = setValues(name, new String[] {value}, append);
+		}
+
+		_mutated = true;
+
+		if ((oldValues != null) && (oldValues.length > 0)) {
+			return oldValues[0];
+		}
+
+		return null;
+	}
+
+	@Override
+	public String[] setValues(String name, String... values) {
+		if (name == null) {
+			throw new IllegalArgumentException();
+		}
+
+		if (values == null) {
```

Comment:
@shuyangzhou You are probably correct. But since we're currently passing 100% of the Portlet 3.0 TCK and I would prefer to add this to my TODO list and send you a subsequent pull request. OK with you?